server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -192,6 +197,72 @@ public Weight createWeight(Query query, ScoreMode scoreMode, float boost) throws
         }
     }
 
+    @Override
+    protected LeafSlice[] slices(List<LeafReaderContext> leaves) {
+        return computeSlices(leaves, queueSizeBasedExecutor.threadPoolExecutor.getPoolSize(), MINIMUM_DOCS_PER_SLICE);
```

Comment:
The current issue says: 
```
A maximum number of slices. The number of slices that a shard level search creates should be capped. To avoid creating more slices than available search threads.
```
So I am reading `available search threads` as the pool size regardless if the thread is already executing another task.

Reply:
to me available == active , but I may be wrong. On the other hand, I am not sure if we will after all tie the number of slices to the number of threads. Maybe for now we should just cap it to something like 10, and later we can refine it.

Reply:
I think we should use `ThreadPoolExecutor#getPoolSize(...)` as basis to determine the `maxSliceNum` here. I think `ThreadPoolExecutor#getActiveCount(...)` can vary quickly over time and I think it is ok of a thread executes multiple slice searches? At the time we attempt to fork we really know whether there is a search thread available to perform a slice search. But basing the number of slices on `ThreadPoolExecutor#getActiveCount(...)` makes `maxSliceNum` not predictable. The `maxSliceNum` is a maximum, but I think it is good to let it be based on the search thread pool size (size that is based on the number cores) compared to a static value. Maybe we should `maxSliceNum` based on `ThreadPoolExecutor#getPoolSize(...) / 2` to always be conservative on the maximum number of slices.

Reply:
I aggre that maybe limited down to half of the pool size might avoid greedy searches using all the pool.

Reply:
It would be a feature to me that a single query could use the entire threadpool rather than a bug. That said, I still like the idea of bounding the number of slices by the size of the threadpool as a way to limit memory usage on small nodes. E.g. if a node has 3 search threads, it is questionable whether 5 slices would do more hard than good as it would certainly increase memory usage (since each slice is collected independently) while barely helping with latency compared to 3 slices?

Reply:
One question that I have is, what happens in a CachedThreadPool case here. Will it return '1' at first and given that it will never use more in the unlucky case it won't maximize the nodes capabilities? Maybe we need to heuristic based on the `getMaxPoolSize()` to scale it well? I haven't looked into what all the flavors return here but if we wanna maximize throughput on idle nodes we might need to look into that?

Reply:
Yes, I think that we will revisit this anyways once we have a thread pool to execute these on. On it, will have a PR up soon. The idea is that the executor will throttle execution of tasks to a certain max concurrent tasks for each IndexSearcher#search call.Hopefully that max (very likely to be getMaxPoolSize) will align with the max number of slices that are available.